pkg/provider/azure_vmssflex_cache.go

```diff
@@ -0,0 +1,302 @@
+		if rerr != nil {
+			klog.Errorf("ListVmssFlexVMsWithOnlyInstanceView failed: %v", rerr)
+			return nil, rerr.Error()
+		}
+
+		for i := range vms {
+			vm := vms[i]
+			if vm.Name != nil {
+				localCache.Store(*vm.Name, &vm)
+				fs.vmssFlexVMNameToVmssID.Store(*vm.Name, key)
+			}
+		}
+
+		return localCache, nil
+	}
+
+	if fs.Config.VmssFlexVMStatusCacheTTLInSeconds == 0 {
+		fs.Config.VmssFlexVMStatusCacheTTLInSeconds = consts.VmssFlexVMStatusCacheTTLInSeconds
+	}
+	return azcache.NewTimedcache(time.Duration(fs.Config.VmssFlexVMStatusCacheTTLInSeconds)*time.Second, getter)
+}
+
+func (fs *FlexScaleSet) getNodeVmssFlexID(nodeName string) (string, error) {
+	fs.lockMap.LockEntry(consts.GetNodeVmssFlexIDLockKey)
+	defer fs.lockMap.UnlockEntry(consts.GetNodeVmssFlexIDLockKey)
+	vmssFlexID, isCached := fs.vmssFlexVMNameToVmssID.Load(nodeName)
+	if !isCached {
+		klog.V(12).Infof("nodeName %s is not saved in vmssFlexVMnameToVmssID map, send a GET request to retrieve its VmssID", nodeName)
+		machine, err := fs.getVirtualMachine(types.NodeName(nodeName), azcache.CacheReadTypeUnsafe)
+		if err != nil {
+			return "", err
+		}
+		vmssFlexID = to.String(machine.VirtualMachineScaleSet.ID)
+		if vmssFlexID == "" {
+			return "", ErrorVmssIDIsEmpty
+		}
+		fs.vmssFlexVMNameToVmssID.Store(nodeName, vmssFlexID)
+		_, err = fs.vmssFlexVMCache.Get(fmt.Sprintf("%v", vmssFlexID), azcache.CacheReadTypeForceRefresh)
+		if err != nil {
+			return "", err
+		}
+		_, err = fs.vmssFlexVMStatusCache.Get(fmt.Sprintf("%v", vmssFlexID), azcache.CacheReadTypeForceRefresh)
+		if err != nil {
+			return "", err
+		}
+
+	}
+	return fmt.Sprintf("%v", vmssFlexID), nil
+}
+
+func (fs *FlexScaleSet) getVmssFlexVMWithoutInstanceView(nodeName string, crt azcache.AzureCacheReadType) (vm compute.VirtualMachine, err error) {
+	vmssFlexID, err := fs.getNodeVmssFlexID(nodeName)
+	if err != nil {
+		return vm, err
+	}
+
+	cached, err := fs.vmssFlexVMCache.Get(vmssFlexID, crt)
+	if err != nil {
+		return vm, err
+	}
+
+	vmMap := cached.(*sync.Map)
+	cachvmedVM, ok := vmMap.Load(nodeName)
+	if !ok {
+		klog.V(2).Infof("did not find node (%s) in the existing cache, which means it is deleted...", nodeName)
+		return vm, cloudprovider.InstanceNotFound
+	}
+
+	return *(cachvmedVM.(*compute.VirtualMachine)), nil
+}
+
+func (fs *FlexScaleSet) getVmssFlexVMStatus(nodeName string, crt azcache.AzureCacheReadType) (vm compute.VirtualMachine, err error) {
+	vmssFlexID, err := fs.getNodeVmssFlexID(nodeName)
+	if err != nil {
+		return vm, err
+	}
+
+	cached, err := fs.vmssFlexVMStatusCache.Get(vmssFlexID, crt)
+	if err != nil {
+		return vm, err
+	}
+
+	vmMap := cached.(*sync.Map)
+	cachvmedVM, ok := vmMap.Load(nodeName)
+	if !ok {
+		klog.V(2).Infof("did not find node (%s) in the existing cache, which means it is deleted...", nodeName)
+		return vm, cloudprovider.InstanceNotFound
+	}
+
+	return *(cachvmedVM.(*compute.VirtualMachine)), nil
```

Comment:
this will not panic in this PR since cachvmedVM wont' be nil, while in long term, it's better check the nil value, that would be safer.

Reply:
I agree the principle in general, but in this current code, if cachvmedVM is nil, it goes into the if !ok logic block. Other existing code like VMSS Uniform does not do this check as well: https://github.com/kubernetes-sigs/cloud-provider-azure/blob/master/pkg/provider/azure_vmss.go#L126